packages/style-engine/class-wp-style-engine-processor.php

```diff
@@ -65,29 +73,40 @@ public function add_rules( $css_rules ) {
 			}
 
 			foreach ( $css_rules as $rule ) {
-				$selector = $rule->get_selector();
-				$at_rule  = $rule->get_at_rule();
+				$rules_group = $rule->get_rules_group();
 
-				/**
-				 * If there is an at_rule and it already exists in the css_rules array,
-				 * add the rule to it.
-				 * Otherwise, create a new entry for the at_rule
+				/*
+				 * Merge existing rule and rules groups objects or create new ones.
+				 * Rules groups and rules are stored in separate arrays to allow for
+				 * unique processing.
 				 */
-				if ( ! empty( $at_rule ) ) {
-					if ( isset( $this->css_rules[ "$at_rule $selector" ] ) ) {
-						$this->css_rules[ "$at_rule $selector" ]->add_declarations( $rule->get_declarations() );
-						continue;
+				if ( $rule instanceof WP_Style_Engine_CSS_Rules_Group ) {
+					if ( isset( $this->css_rules_groups[ $rules_group ] ) ) {
+						$this->css_rules_groups[ $rules_group ]->add_rules( $rule->get_rules() );
+					} else {
+						$this->css_rules_groups[ $rules_group ] = $rule;
 					}
-					$this->css_rules[ "$at_rule $selector" ] = $rule;
 					continue;
 				}
 
-				// If the selector already exists, add the declarations to it.
-				if ( isset( $this->css_rules[ $selector ] ) ) {
-					$this->css_rules[ $selector ]->add_declarations( $rule->get_declarations() );
-					continue;
+				if ( $rule instanceof WP_Style_Engine_CSS_Rule ) {
```

Comment:
If this change is too heavy, we can revert and just introduce the `WP_Style_Engine_CSS_Rules_Group` class to allow for separate processing.

Reply:
What would the alternative look like? Just trying to wrap my head around the proposal and what the trade-offs might be.

Reply:
Reverting would just lose the ability to group nested rules. 

Not a huge deal.